### PR TITLE
docs: rfc for table compaction

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add RFC for table compaction.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: docs/rfcs/2023-02-01-table-compaction.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    fn pick(
+        &self,
+        ctx: CompactionContext,
+        levels: &LevelMetas,
+    ) -> Result<CompactionTask>;
+}
+```
+
+The most suitable compaction strategy for time-series scenario would be 
+a hybrid strategy that combines time window compaction with size-tired compaction, just like [Cassandra](https://cassandra.apache.org/doc/latest/cassandra/operating/compaction/twcs.html) and [ScyllaDB](https://docs.scylladb.com/stable/architecture/compaction/compaction-strategies.html#time-window-compaction-strategy-twcs) does.
```
Comment (evenyag) [inline docs/rfcs/2023-02-01-table-compaction.md:78]: I think we could only implement the size-tired compaction first.

File: docs/rfcs/2023-02-01-table-compaction.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+But currently it only supports level 0. SST files in level 0 does not guarantee to contain only rows with disjoint time ranges. 
+That is to say, different SST files in level 0 may contain overlapped timestamps. 
+The consequence is, in order to retrieve rows in some time range, all files need to be scanned, which bring a lot of IO overhead.
+
+Also, just like other LSMT engines, delete/update to existing primary keys are converted to new rows with delete/update mark and appended to SSTs on flushing. 
+We need to merge the operations to same primary keys so that we don't have to go through all SST files to find the final state of these primary keys.  
+
+## Goal
+
+Implement a compaction mechanism to:
```
Comment (killme2008) [inline docs/rfcs/2023-02-01-table-compaction.md:24]: Another goal is implementing a compaction framework that may be used in data rollup/indexing rewrite tasks etc. in future.

File: docs/rfcs/2023-02-01-table-compaction.md (hunk 1 above)
Comment (killme2008) [reply docs/rfcs/2023-02-01-table-compaction.md:78]: Agree, getting started from a simple one.
I've implemented STCS(SizeTieredCompactionStrategy) in ceresdb before, I can help to do this work.

File: docs/rfcs/2023-02-01-table-compaction.md (hunk 2 above)
Comment (waynexia) [reply docs/rfcs/2023-02-01-table-compaction.md:24]: Yes, it's hard to build efficient index in L0, which doesn't contain many useful information IMO. So compaction is a kind of pre-requirement of that I think

File: docs/rfcs/2023-02-01-table-compaction.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    ) -> Result<CompactionTask>;
+}
+```
+
+The most suitable compaction strategy for time-series scenario would be 
+a hybrid strategy that combines time window compaction with size-tired compaction, just like [Cassandra](https://cassandra.apache.org/doc/latest/cassandra/operating/compaction/twcs.html) and [ScyllaDB](https://docs.scylladb.com/stable/architecture/compaction/compaction-strategies.html#time-window-compaction-strategy-twcs) does.
+
+We can first group SSTs in level n into buckets according to some predefined time window. Within that window, 
+SSTs are compacted in a size-tired manner (find SSTs with similar size and compact them to level n+1). 
+SSTs from different time window are neven compacted together.
```
Comment (waynexia) [inline docs/rfcs/2023-02-01-table-compaction.md:82]:
```suggestion
SSTs from different time windows are neven compacted together.
```

**Review by evenyag — APPROVED** [review]

**Review by waynexia — APPROVED** [review]

**Pull request merged**